src/Microsoft.SqlTools.ServiceLayer/LanguageExtensibility/ExternalLanguageOperations.cs

```diff
@@ -0,0 +1,359 @@
+                }
+                return new List<ExternalLanguage>(dic.Values);
+            }
+            catch
+            {
+                throw;
+            }
+        }
+
+        private string GetCreateScript(ExternalLanguage language, Dictionary<string, object> parameters)
+        {
+            return GetLanguageModifyScript(language, language.Contents, parameters, ModifyType.Create);
+        }
+
+        private string GetUpdateScript(ExternalLanguage language, ExternalLanguageContent content, Dictionary<string, object> parameters, ContentModifyType contentModifyType)
+        {
+            return GetLanguageModifyScript(language, new List<ExternalLanguageContent> { content }, parameters, ModifyType.Alter, contentModifyType);
+        }
+
+        private string GetLanguageModifyScript(ExternalLanguage language, List<ExternalLanguageContent> contents, Dictionary<string, object> parameters, ModifyType modifyType, ContentModifyType contentModifyType = ContentModifyType.Add)
+        {
+            string contentScript = string.Empty;
+            for (int i = 0; i < contents.Count; i++)
+            {
+                var content = contents[i];
+                string seperator = contentScript == string.Empty ? "" : ",";
+                contentScript = $"{contentScript}{seperator}{GetLanguageContent(content, i, parameters)}";
+            }
+
+            string ownerScript = string.IsNullOrWhiteSpace(language.Owner) ? "" : $"AUTHORIZATION {language.Owner}";
+            string scriptAction = modifyType == ModifyType.Create ? CreateScript : AlterScript;
+            string contentAction = "FROM";
+            if (modifyType == ModifyType.Alter)
+            {
+                switch (contentModifyType)
+                {
+                    case ContentModifyType.Modify:
+                        contentAction = SetContentScript;
+                        break;
+                    case ContentModifyType.Add:
+                        contentAction = AddContentScript;
+                        break;
+                    case ContentModifyType.Remove:
+                        contentAction = RemoveContentScript;
+                        break;
+                }
+            }
+            return $@"
+{scriptAction} {language.Name}
+{ownerScript}
+{contentAction} {contentScript}
+";
+        }
+
+        private string AddStringParameter(string paramName, string prefix, string postfix, string paramValue, Dictionary<string, object> parameters)
+        {
+            string script = string.Empty;
+            string sqlParam = $"@{paramName}{postfix}";
+            /*
+            if (!string.IsNullOrWhiteSpace(paramValue))
+            {
+                script = $", {paramName} = N'{sqlParam}'";
+                parameters.Add($"{sqlParam}", paramValue);
+            }
+            */
+
+            if (!string.IsNullOrWhiteSpace(paramValue))
+            {
+                script = $"{prefix} {paramName} = N'{paramValue}'";
```

Comment:
This should be escaping ' in the value

Reply:
it's escaped. not sure which one you are talking about N'{paramValue}'

---
In reply to: [374373171](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r374373171) [](ancestors = 374373171)

Reply:
Where is it escaped? If paramValue is a string such as `my'teststring` isn't that going to break the query text?

Reply:
ah ok, should it be this? N'[{paramValue}]'

---
In reply to: [374942464](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r374942464) [](ancestors = 374942464)

Reply:
No, you'd double up the apostrophe's

`paramValue = paramValue.replace(/'/g, '\'\''); `

Although using SqlParameter here would be even better and handles doing that for you.

Reply:
I tried sql param and verified that the tsql doesn't support sql param in this case. looks like sql params are not supported for ddl statements? not sure why when I pass sql params with byte array for content it works but string parameters won't work. anyway will fix the escape problem although because this parameters  are file path and file name, they cannot have quote in the name but still won't hurt to add that. but why wrapping it in bracket won't work? I thought this is how escape works in SQL statements 

---
In reply to: [374950398](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r374950398) [](ancestors = 374950398)

Reply:
Brackets are used to quote object names - not strings. Just try running these commands :

`select N'[test]'`
output is `[test]`

and trying to run `select N'[test'string]'` results in a parse error because it sees the single quote in the string as the ending quote.